Add function to export user report data from Gophish (#33)
cisagov/gophish-tools: README.md

* `gophish-complete` - Completes a campaign in GoPhish and/or outputs a
  GoPhish campaign summary.

> * `gophish-export` - Exports all the data from an assessment within GoPhish

Reviewer: Please make an issue to replace all instances of `GoPhish` with `Gophish` in this project. We've slowly been making the change in other projects as we do updates and I don't want to scope creep this PR.

Author: See [here](https://github.com/cisagov/gophish-tools/issues/35) for the new issue

  into a single JSON file. In addition, user report JSONs for each campaign in

>   an assessment will also be generated, for later import into the PCA database.

Reviewer 2: Remove reference to PCA- this repo should stand on it's own as a Gophish helper utility, regardless of whether or not it is used for PCAs.
```suggestion
  an assessment will also be exported.
```

Author: See 2d96175613954e3c341aa38614d80369113217af

Reviewer 2: Thanks!  FYI, you can use [GitHub's "Commit suggestion" feature](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) to save time when making changes like this in the future.

| interface_type | Type of interface GoPhish will use with mail server. | string | "SMTP" | no |
| ignore_cert | Indicate if GoPhish should ignore certs with mail server. | boolean | `True` | no |

## User Report Field Dictionary ##

> User report JSONs are also exported by ```gophish-export```. User reports

Reviewer 2: Single backticks are sufficient here:
```suggestion
User report JSONs are also exported by `gophish-export`. User reports
```

summarize data from targeted user clicks on phishing emails generated by a

> campaign. These JSON files are later imported into the PCA database.

Reviewer 2: Remove reference to PCA- this repo should stand on it's own as a Gophish helper utility, regardless of whether or not it is used for PCAs.
```suggestion
campaign.
```

Author: See 2d96175613954e3c341aa38614d80369113217af

| Name | Description | Type | Default | Required |
|------|-------------|:----:|:-------:|:--------:|
| assessment | Assessment ID that this user report is associated with. | string | | yes |
| campaign | Campaign ID that this user report is associated with. | string | | yes |
| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as `null`. | string | `null` | no |
| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDThh:mm.ss" | datetime | | yes |

> | total_num_reports | Total number of user reports received during a campaign. | integer | | yes |

Reviewer: `Default`/`Required` doesn't make much sense for a JSON output. I also updated one of the descriptions to reflect a change I suggested in the code.

```suggestion
| Name | Description | Type |
|------|-------------|:----:|
| assessment | Assessment ID that this user report is associated with. | string |
| campaign | Campaign ID that this user report is associated with. | string |
| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as an empty string. | string |
| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDThh:mm.ss" | datetime |
| total_num_reports | Total number of user reports received during a campaign. | integer |
```

Author: Changes incorporated in cf4906f0b51ec10f488d8d013bed0d28003dd4ce

> | first_report | The first report (click) generated by a targeted user. Formatted by YYYY-MM-DDT: hh:mm.ss"| datetime | | yes |

Reviewer 2: Is `YYYY-MM-DDT: hh:mm.ss` really the correct format or should it be `YYYY-MM-DDThh:mm.ss`?

Author: See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

Reviewer 2: It looks like you corrected the format in https://github.com/cisagov/gophish-tools/pull/33/commits/ccdf910c06d009dca6304f8223285cb2d9d99042, but then overwrote it with https://github.com/cisagov/gophish-tools/commit/bac2cb88e00b183bdbae8f990c3ad5a48b01ef05.  I assume that you did not mean to do that.  Please let me know when it's fixed.

Author: Fixed here 71d9fc125456322326ea57a91ce7d86fbf07d45e

Reviewer 2: Thanks!

> | total_num_reports | The total number of user reports received during a campaign.| integer | | no

Reviewer 2: I recommend sorting these rows alphabetically by field name.

Author: I put them in this order because this is the order I have them on in the JSON file. Are you sure you wanna change the field order?

Reviewer 3: The order of the fields in the JSON file shouldn't be relevant, right?  It isn't a CSV.  I think @dav3r is right here.

If part of your code depends on the order of JSON dictionary keys that probably indicates a problem, since JSON does not make any promises about such ordering.  If you parse a JSON and then re-export it the order of the keys can change depending on the hash map implementation being used internally.

Author: See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

Reviewer 2: @JCantu248, https://github.com/cisagov/gophish-tools/pull/33/commits/bac2cb88e00b183bdbae8f990c3ad5a48b01ef05 does not correct the alphabetical order.

Also, please use more descriptive messages in your commits: `Merging remote` is not particularly helpful here.  We strongly encourage everyone to follow the commit message style and guidelines described in [this post](https://chris.beams.io/posts/git-commit/#seven-rules), as referenced in our [Development Guide](https://github.com/cisagov/development-guide#git-github-and-you).

Reviewer 2: Thanks @JCantu248 for fixing the alphabetization in https://github.com/cisagov/gophish-tools/pull/33/commits/cc24b7780b4ff24ca8f43633d2dce6a214d86892. 👍

> | total_num_reports | The total number of user reports received during a campaign.| integer | | no

Reviewer 2: Clean up wording a bit and remove PCA-specific language:
```suggestion
| Name | Description | Type | Default | Required |
|------|-------------|:----:|:-------:|:--------:|
| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as `null`. | string | `null` | no |
| assessment | Assessment ID that this user report is associated with. | string | | yes |
| campaign | Campaign ID that this user report is associated with. | string | | yes |
| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDT: hh:mm.ss" | datetime | | yes |
| total_num_reports | Total number of user reports received during a campaign. | integer | | no |
```

> | total_num_reports | The total number of user reports received during a campaign.| integer | | no

Reviewer 2: I thought over in [this discussion](https://github.com/cisagov/pca-core/pull/5#discussion_r651079451), we were saying that `total_num_reports` cannot be blank, i.e. it is required?

Author: See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

Reviewer 2: It looks like you changed this field to be required in ccdf910, but then overwrote it with bac2cb8. I assume that you did not mean to do that. Please let me know when it's fixed.

Author: See cc24b7780b4ff24ca8f43633d2dce6a214d86892

Reviewer 2: https://github.com/cisagov/gophish-tools/commit/cc24b7780b4ff24ca8f43633d2dce6a214d86892 fixes the alphabetization issue, but `total_num_reports` is still listed as optional.  Did you mean to put this comment in the conversation above?

Author: Nah, it was an oversight. See 51786dc6d5b6d82de55ecbf822644fe2fe43ce71
